designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtypedata/ProcessingTypeDataProvider.scala

```diff
@@ -34,75 +32,126 @@ trait ProcessingTypeDataProvider[+T, +C] {
+  def mapCombined[CC](fun: C => CC): ProcessingTypeDataProvider[T, CC] =
+    new TransformingProcessingTypeDataProvider[T, C, T, CC](this, _.mapCombined(fun))
 
-      override def all(implicit user: LoggedUser): Map[ProcessingType, Y] =
-        ProcessingTypeDataProvider.this.all.mapValuesNow(fun)
+}
 
-      override def combined: C = ProcessingTypeDataProvider.this.combined
-    }
+class TransformingProcessingTypeDataProvider[T, C, TT, CC](
+    observed: ProcessingTypeDataProvider[T, C],
+    transformState: ProcessingTypeDataState[T, C] => ProcessingTypeDataState[TT, CC]
+) extends ProcessingTypeDataProvider[TT, CC] {
 
+  private var stateValue: ProcessingTypeDataState[TT, CC] = transformState(observed.state)
+
+  override private[processingtypedata] def state: ProcessingTypeDataState[TT, CC] = synchronized {
+    val currentObservedState = observed.state
+    if (currentObservedState.stateIdentity != stateValue.stateIdentity) {
+      stateValue = transformState(currentObservedState)
+    }
+    stateValue
   }
 
-  def mapCombined[CC](fun: (=> C) => CC): ProcessingTypeDataProvider[T, CC] = {
+}
+
+object ProcessingTypeDataProvider {
 
-    new ProcessingTypeDataProvider[T, CC] {
+  val noCombinedDataFun: () => Nothing = () =>
+    throw new IllegalStateException(
+      "Processing type data provider does not have combined data!"
+    )
+
+  def apply[T, C](stateValue: ProcessingTypeDataState[T, C]): ProcessingTypeDataProvider[T, C] =
+    new ProcessingTypeDataProvider[T, C] {
+      override private[processingtypedata] def state: ProcessingTypeDataState[T, C] = stateValue
+    }
 
-      override def forType(processingType: ProcessingType)(implicit user: LoggedUser): Option[T] =
-        ProcessingTypeDataProvider.this.forType(processingType)
+  def apply[T, C](
+      allValues: Map[ProcessingType, ValueWithPermission[T]],
+      combinedValue: C
+  ): ProcessingTypeDataProvider[T, C] =
+    new ProcessingTypeDataProvider[T, C] {
 
-      override def all(implicit user: LoggedUser): Map[ProcessingType, T] = ProcessingTypeDataProvider.this.all
+      override private[processingtypedata] val state: ProcessingTypeDataState[T, C] = ProcessingTypeDataState(
+        allValues,
+        () => combinedValue,
+        allValues
+      )
 
-      override def combined: CC = fun(ProcessingTypeDataProvider.this.combined)
     }
 
-  }
+  def withEmptyCombinedData[T](
+      allValues: Map[ProcessingType, ValueWithPermission[T]]
+  ): ProcessingTypeDataProvider[T, Nothing] =
+    new ProcessingTypeDataProvider[T, Nothing] {
+
+      override private[processingtypedata] val state: ProcessingTypeDataState[T, Nothing] = ProcessingTypeDataState(
+        allValues,
+        noCombinedDataFun,
+        allValues
+      )
+
+    }
 
 }
 
-class MapBasedProcessingTypeDataProvider[T, C](map: Map[ProcessingType, ValueWithPermission[T]], getCombined: => C)
-    extends ProcessingTypeDataProvider[T, C] {
+trait ProcessingTypeDataState[+T, +C] {
```

Comment:
IMO this name is also too generic. We have `ValueWithPermission` in the interface and it's not obvious for me why it's there and how it relates to the State.

Reply:
With regards to `ValueWithPermission`: Due to the fact, that `ProcessingType's` concept is deeply rooted in our application, you can see it in many tests. Because of that, we use `ProcessingTypeDataProvider.apply(Map)` instead of creating reloadable `ProcessingTypeDataReload`, and we specify that any user will have an access to given `ProcessingType` Data, not only those with `AccessRightsToCategory`. Also, we wanted to have a backward compatibility to previous `categoriesConfig` (see `LegacyConfigProcessCategoryService` and notice next to `CategoryConfig`), so I decided to use `AnyUser`access also for the legacy config. It would be difficult to check categories access using the old format, because `ProcessCategoryService` is inside `CombinedProcessingTypeData`. We can discuss it offline. Maybe we should break compatibility and leave `AnyUser` access only for tests purpose?

With regards to `ProcessingTypeDataState` name. I have no cloud, how to call it. I need it to provide caching. It keeps a state (value) that is cached and restricted by `ProcessingTypeDataProvider`.